sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala

```diff
@@ -0,0 +1,204 @@
+      finishTimestamp - startTimestamp
+    }
+  }
+}
+
+private[sql] class ThriftServerUIEventListener(val conf: SparkConf)
+  extends ThriftServerEventListener with SparkListener {
+
+  import ThriftServerUIEventListener._
+
+  var sessionList = new HashMap[SessionHandle, SessionInfo]
+  var executeList = new HashMap[String, ExecutionInfo]
+  val retainedStatements =
+    conf.getInt("spark.thriftserver.ui.retainedStatements", DEFAULT_RETAINED_STATEMENTS)
+  val retainedSessions =
+    conf.getInt("spark.thriftserver.ui.retainedSessions", DEFAULT_RETAINED_SESSIONS)
+  var totalRunning = 0
+
+  override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
+    val jobGroup = for (
+      props <- Option(jobStart.properties);
+      statement <- Option(props.getProperty(SparkContext.SPARK_JOB_DESCRIPTION))
+    ) yield statement
+
+    jobGroup match {
+      case Some(statement: String) => {
+        val ret = executeList.find( _ match {
+          case (id: String, info: ExecutionInfo) => {
+            info.statement == statement
+          }
+        })
+        if(ret.isDefined) {
+          ret.get._2.jobId = jobStart.jobId.toString
+          ret.get._2.groupId = jobStart.properties.getProperty(SparkContext.SPARK_JOB_GROUP_ID,"")
+        }
+      }
+    }
+  }
+
+  override def onSessionCreated(ip: String, session: HiveSession): Unit = {
+    val info = new SessionInfo(session, System.currentTimeMillis(), ip)
+    sessionList(session.getSessionHandle) = info
+    trimSessionIfNecessary()
+  }
+
+  override def onSessionClosed(session: HiveSession): Unit = {
+    sessionList(session.getSessionHandle).finishTimestamp = System.currentTimeMillis()
+  }
+
+  override def onStatementStart(id: String, session: HiveSession, statement: String): Unit = {
+    val info = new ExecutionInfo(statement, session, System.currentTimeMillis())
+    info.state = ExecutionState.STARTED
+    executeList(id) = info
+    trimExecutionIfNecessary()
+    sessionList(session.getSessionHandle).totalExecute += 1
+    totalRunning += 1
+  }
+
+  override def onStatementParse(id: String, executePlan: String): Unit = {
+    executeList(id).executePlan = executePlan
+    executeList(id).state = ExecutionState.COMPILED
+  }
+
+  override def onStatementError(id: String, errorMessage: String, errorTrace: String): Unit = {
+    executeList(id).finishTimestamp = System.currentTimeMillis()
+    executeList(id).detail = errorMessage
+    executeList(id).state = ExecutionState.FAILED
+    totalRunning -= 1
+  }
+
+  override def onStatementFinish(id: String): Unit = {
+    executeList(id).finishTimestamp = System.currentTimeMillis()
+    executeList(id).state = ExecutionState.FINISHED
+    totalRunning -= 1
+  }
+
+  private def trimExecutionIfNecessary() = synchronized {
+    if (executeList.size > retainedStatements) {
+      val toRemove = math.max(retainedStatements / 10, 1)
+      executeList.toList.sortWith(compareExecutionDesc).take(toRemove).foreach { s =>
```

Comment:
I think you can use `sortBy(_._2.startTimestamp)` instead of having to define a custom comparison function.